### PR TITLE
simplify code contributions + reviews by fully automating the dev setup.

### PR DESCRIPTION
**Pull request by nisarhassan12**
Hi :slightly_smiling_face:

This Pr simplifies code contributions by fully automating the dev setup with Gitpod(a full-featured online cloud IDE) which is free for OSS. With a single click, it'll launch a workspace and automatically:

- clone the dvc.org repo.
- install the dependencies.
- run `yarn build`.
- start the webserver via `yarn start`.

so that anyone interested in contributing can start straight away without any friction.

this is how it looks:

image

You can give it a try on my fork of the repo via the following link:

https://gitpod.io/#https://github.com/nisarhassan12/dvc.org

A lot of popular OSS projects out there are already using Gitpod to streamline the experience for their contributors and maintainers, some of them are: 

- https://github.com/freeCodeCamp/freeCodeCamp
-  https://github.com/facebook/docusaurus
-  https://github.com/mozilla/pdf.js/

You can find a brief list of them at https://contribute.dev

image

Prs can also be reviewed from within Gitpod.

### CODE REVIEW

**Comment by shcheklein** [comment]
@nisarhassan12 looks cool!

@casperdcl link check fails again, looks like a false positive.

**Comment by casperdcl** [comment]
much sneaky such wow. This pattern:

```
[](https://gitpod.io/#https://github.com/iterative/dvc.org)
```

is pretty darn hard to fix. I'd recommend just adding it to the excludes rather than coming up with a solution. Currently the above line results in:

```bash
https://github.com/iterative/dvc.org)
https://gitpod.io/#https://github.com/iterative/dvc.org
```

so `https://github.com/iterative/dvc.org)` would have to be added to `scripts/exclude-links.txt`

**Comment by shcheklein** [comment]
why does it result in https://github.com/iterative/dvc.org) ... don't we ignore `)`?

**Comment by casperdcl** [comment]
nope because parentheses are valid chars in URLs. There are many in the docs, e.g. https://en.wikipedia.org/wiki/Hyperparameter_(machine_learning)

**Comment by casperdcl** [comment]
I'll suggest a minor fix in a PR but would be best to double-check...

**Comment by shcheklein** [comment]
@nisarhassan12 could please the latest master? to fix CI check

File: .gitpod.yml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+tasks:
+  - init: yarn install && yarn build
+    command: yarn develop
+
+ports:
+  - port: 8000
+    onOpen: open-preview
```
Comment (casperdcl) [inline .gitpod.yml:7]: does this file have to be in the repo root? Couldn't find anything in https://www.gitpod.io/docs/config-gitpod-file/ about location

Comment (nisarhassan12) [reply .gitpod.yml:7]: Yeah, it has to be in the repo root. Sorry! that It's not documented.

File: content/docs/user-guide/contributing/docs.md [hunk 2]
```diff
@@ -51,6 +51,14 @@ Otherwise, please refer to the following procedure:
 
 We will review your PR as soon as possible. Thank you for contributing!
 
+## Automated development environment in a single click
```
Comment (shcheklein) [inline content/docs/user-guide/contributing/docs.md:54]: let's make it a paragraph in the Dev environment. Like a second paragraph for example should be a good fit. May remove "locally" from the first paragraph.

File: .gitpod.yml (hunk 1 above)
Comment (shcheklein) [reply .gitpod.yml:7]: 👍

**Comment by nisarhassan12** [comment]
> @nisarhassan12 could you please rebase against or merge the latest master? to fix CI check

Thanks! @shcheklein I have rebased it againist the latest master.

**Comment by shcheklein** [comment]
@nisarhassan12 awesome, one more ask - please check the comment. I think we can make it part of the dev env section in the docs.

File: content/docs/user-guide/contributing/docs.md (hunk 2 above)
Comment (nisarhassan12) [reply content/docs/user-guide/contributing/docs.md]: Thanks! I have made the change.

**Comment by shcheklein** [comment]
done, merged! and I moved a bit in the master ... thanks 🙏 Hopefully we'll be getting even more contributors now. Will you make a PR to the list of sites, btw?

**Comment by nisarhassan12** [comment]
@shcheklein  sure! please open an issue here https://github.com/gitpod-io/contribute.dev/ :slightly_smiling_face:

**Comment by jorgeorpinel** [comment]
Sorry for the delay. This is really cool (:

The button looks a little strange in https://dvc.org/doc/user-guide/contributing/docs#development-environment though: 

image

☝️ I can address this in a future PR, but it's not working for me:

image

Can you maybe check what the problem is, @nisarhassan12? It works from https://contribute.dev/

**Comment by shcheklein** [comment]
Yep, does not work for me either (worked yesterday).

Also, we need to adjust CSS a bit to remove the "external link" icon. @jorgeorpinel mind creating a ticket for this?

**Comment by nisarhassan12** [comment]
> Can you maybe check what the problem is, @nisarhassan12? It works from https://contribute.dev/

Sure. I will shortly make a Pr with the fix.

**Comment by jorgeorpinel** [comment]
Thanks again Nisar!

**Comment by jorgeorpinel** [comment]
Hmmm I tried using Gitpod to format the note I mentioned above. It was all working nicely, I even installed a VS Code plugin to the IDE. But after committing with Git, I'm unable to push because I didn't grant Gitpod permission to push directly to dvc.org

image

Open source contributors will never have this permission. Can it be made so a fork is created in their accounts first, and then Gitpod launched from there? Or maybe if I was using a different user, that would've been the behavior... Not sure

**Comment by jorgeorpinel** [comment]
> Or maybe if I was using a different user, that would've been the behavior... Not sure

Nope. I tried with @AyokVoluntarios whici I also own and it tries to use the upstream directly:

image

So even if I tell Github to let Gitpod push to my repos, I simply can't push to iterative/dvc.org from that account.

image

The user would need to configure the Git remote and branch manually from the IDE. Do you know whether that's the only solution, @nisarhassan12? Thanks

**Comment by shcheklein** [comment]
@jorgeorpinel good catch, let's create a separate ticket and ping @nisarhassan12 there?

**Comment by jorgeorpinel** [comment]
Sure, done: #1253

**Pull request merged**